Content.Shared/Materials/SharedMaterialStorageSystem.cs

```diff
@@ -0,0 +1,227 @@
+﻿using System.Linq;
+using Content.Shared.Interaction;
+using Content.Shared.Stacks;
+using Robust.Shared.GameStates;
+
+namespace Content.Shared.Materials;
+
+/// <summary>
+/// This handles storing materials and modifying their amounts
+/// <see cref="MaterialStorageComponent"/>
+/// </summary>
+public abstract class SharedMaterialStorageSystem : EntitySystem
+{
+    /// <inheritdoc/>
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<MaterialStorageComponent, InteractUsingEvent>(OnInteractUsing);
+        SubscribeLocalEvent<MaterialStorageComponent, ComponentGetState>(OnGetState);
+        SubscribeLocalEvent<MaterialStorageComponent, ComponentHandleState>(OnHandleState);
+    }
+
+    private readonly List<MaterialStorageComponent> _defferedMaterialChangeEvent = new();
+
+    private void OnGetState(EntityUid uid, MaterialStorageComponent component, ref ComponentGetState args)
+    {
+        args.State = new MaterialStorageComponentState(component.Storage, component.MaterialWhiteList);
+    }
+
+    private void OnHandleState(EntityUid uid, MaterialStorageComponent component, ref ComponentHandleState args)
+    {
+        if (args.Current is not MaterialStorageComponentState state)
+            return;
+
+        component.Storage = state.Storage;
+        component.MaterialWhiteList = state.MaterialWhitelist;
```

Comment:
This will store refs to the state data which means it won't be handled correctly for prediction, you need to copy the data (using linq is fine). I often get caught out by this and electro has to point it out.